pinot-core/src/main/java/org/apache/pinot/core/plan/SelectionPlanNode.java

```diff
@@ -59,12 +59,15 @@ public SelectionPlanNode(IndexSegment indexSegment, QueryContext queryContext) {
             Math.min(limit, DocIdSetPlanNode.MAX_DOC_PER_CALL)).run();
         return new SelectionOnlyOperator(_indexSegment, _queryContext, expressions, transformOperator);
       } else {
-        // Selection order-by
-        if (orderByExpressions.size() == expressions.size()) {
+        if (isAllOrderByColumnsSorted(orderByExpressions)) { // no sorting needed
+          TransformOperator transformOperator =
+                  new TransformPlanNode(_indexSegment, _queryContext, expressions, DocIdSetPlanNode.MAX_DOC_PER_CALL).run();
```

Comment:
We can reduce the block size in `TransformOperator`
```suggestion
                  new TransformPlanNode(_indexSegment, _queryContext, expressions, Math.min(limit + _queryContext.getOffset(), DocIdSetPlanNode.MAX_DOC_PER_CALL))).run();
```